### PR TITLE
doc: Add RPC interface guidelines

### PR DESCRIPTION
**Pull request by laanwj**
Could be useful. Anything that I forgot?

### CODE REVIEW

File: doc/developer-notes.md [hunk 1]
```diff
@@ -495,3 +495,68 @@ Git and GitHub tips
...
+    new ones.
+
+- Try not to overload methods on argument type. E.g. don't make `getblock(true)` and `getblock("hash")`
+  do different things.
+
+  - *Rationale*: This is impossible to use with `bitcoin-cli`, and can be surprising to users.
+
+  - *Exception*: Some RPC calls can take both an `int` and `bool`, most notably when a bool was switched
+    to a multi-value, or due to other historical reasons. **Always** have false map to 0 and
+    true to 1 in this case.
```
Comment (kallewoof) [inline doc/developer-notes.md]: This is not the case right now. At the moment, 0 maps to false and !0 maps to true. See e.g. `getrawtransaction`'s `verbose` argument handling.

File: doc/developer-notes.md [hunk 2]
```diff
@@ -495,3 +495,68 @@ Git and GitHub tips
...
+
+- Don't forget to fill in the argument names correctly in the RPC command table.
+
+  - *Rationale*: If not, the call can not be used with name-based arguments.
+
+- Set okSafeMode in the RPC command table to a sensible value: safe mode is when the
+  block chain is regarded to be in a confused state, and the client deems it unsafe to
+  do anything irreversible such as send. Anything that just queries should be permitted.
```
Comment (kallewoof) [inline doc/developer-notes.md]: Thanks for this! I vaguely presumed this was the case but never found anything explicitly stating it.

**Review by kallewoof — COMMENTED** [review]
Nice.

**Comment by jonasschnelli** [comment]
Great addition.
Concept ACK.

File: doc/developer-notes.md [hunk 3]
```diff
@@ -495,3 +495,68 @@ Git and GitHub tips
...
+  - *Exception*: Some RPC calls can take both an `int` and `bool`, most notably when a bool was switched
+    to a multi-value, or due to other historical reasons. **Always** have false map to 0 and
+    true to 1 in this case.
+
+- Don't forget to fill in the argument names correctly in the RPC command table.
+
+  - *Rationale*: If not, the call can not be used with name-based arguments.
+
+- Set okSafeMode in the RPC command table to a sensible value: safe mode is when the
+  block chain is regarded to be in a confused state, and the client deems it unsafe to
```
Comment (fanquake) [inline doc/developer-notes.md]: nit: s/block chain/blockchain

File: doc/developer-notes.md (hunk 1 above)
Comment (laanwj) [inline doc/developer-notes.md]: EH, I have that the wrong way around, thanks.

Comment (laanwj) [inline doc/developer-notes.md]: Or not. What I meant here is:
```
false=0
true=1
```
E.g, the C++ mapping, not the bash mapping. 
If this wording is confusing, can you suggest another way.

Comment (kallewoof) [inline doc/developer-notes.md]: I read it as "interpret 0 as false and 1 as true, and anything else is [??]".

I.e. if a user passes 2 or -1 to the new RPC command which maps as you said, how will it treat values beyond [0, 1]?

Am I misunderstanding?

Comment (laanwj) [inline doc/developer-notes.md]: I don't say anything about values besides 0 and 1 on purpose. Those may either map to 1, or to something else entirely (referring to "when bool has switched to multi-value").

Comment (kallewoof) [inline doc/developer-notes.md]: Ohh I was reading this backwards. You're talking about when a bool becomes an int, and I about when an int is interpreted as a bool. I'll think on wording to make this obvious.

Comment (kallewoof) [inline doc/developer-notes.md]: Re-reading this, I think I was simply not reading it properly. I think it looks fine as is.

**Comment by jnewbery** [comment]
Looks good. ACK 359f71d0fb243fa49884132c0fb0c24118d77842

**Comment by sipa** [comment]
ACK

**Pull request merged**